docs(toc): add table of contents

JonathanPaugh opened the pull request:
## Description

Adds a table of contents section to the right side of the content area in the documentation. This new section contains the subheadings that were previously on the left sidebar. The toc section will responsively nest itself inside the page contents below the header for mobile displays.

Desktop:
![82987AD5-D543-4F4F-8F74-65544EF70F3E](https://user-images.githubusercontent.com/33143001/173467972-33b0d28b-d1ba-4ab8-ab12-dcf6f7716c0f.jpeg)

Mobile:
<img width="602" alt="image" src="https://user-images.githubusercontent.com/33143001/173468066-069f716c-4ff2-4244-979f-fea68c36aba5.png">


## Pull Request Checklist

 - [x] Ask the contributors if a similar effort is already in process or has been solved.
 - [x] Review the [contribution guidelines](https://github.com/dialpad/dialtone/blob/staging/.github/CONTRIBUTING.md).
 - [x] Use `staging` as your pull request's base branch. (All PRs using `production` as its base branch will be declined).
 - [x] Ensure all `gulp` scripts successfully compile.
 - [x] Update, remove, or extend all affected documentation.
 - [x] Ensure no private Dialpad links or info are in the code or pull request description (Dialtone is a public repo!).

## Obligatory GIF (super important!)
![](https://c.tenor.com/q9mxm8_8lG4AAAAd/git-merge-git.gif)

JonathanPaugh commented on docs/.vuepress/config.ts:
```diff
@@ -81,9 +83,11 @@ export default defineUserConfig({
     code: {
       lineNumbers: false
+    },
+    extractHeaders: {
+      level: [2] // Generated data header levels (used for toc)
     }
```
Toc header depth can be set here, currently only renders level 2 sub-headers that were in the sidebar.

JonathanPaugh commented on docs/.vuepress/theme/layouts/Layout.vue:
```diff
@@ -1,25 +1,101 @@
-            <icon-launch class="d-svg--size16" />
-          </a>
+    <template #page>
+      <Home v-if="$frontmatter.home" />
+        <div v-else class="main-content">
+          <Transition
+            name="fade-slide-y"
+            mode="out-in"
+            @before-enter="onBeforeEnter"
+            @before-leave="onBeforeLeave"
+          >
+            <Page :key="$page.path">
+              <template #top>
+                <PageHeader :path="$page.path" :frontmatter="$frontmatter">
+                  <template #content-bottom>
+                    <PageToc v-if="isMobile" :headers="$page.headers" />
+                  </template>
+                </PageHeader>
+              </template>
+            </Page>
+          </Transition>
+          <PageToc v-if="!isMobile" :headers="$page.headers" class="d-pr32 d-pt24" />
         </div>
-      </section>
     </template>
```
The main layout structure was modified to allow the toc to work with the page content.

Instead of inserting the header into the `#page-top` slot, the `#page` slot is now overridden.
This allows wrapping the `Page` and `PageToc` components in a flexbox so they display side-by-side.

JonathanPaugh commented on docs/.vuepress/theme/layouts/Layout.vue:
```diff
@@ -1,25 +1,101 @@
+    z-index: 10;
+  }
+
+  @media screen and (max-width: 980px) {
+    .main-content {
+      flex-wrap: wrap;
+      flex-direction: column-reverse;
+    }
```
I ended up using a breakpoint instead of the content width break implementation so the location of the `Toc` component in DOM can also be moved at this breakpoint. Also, this will break at the same exact point at the sidebar (980px).

josedialpad commented on docs/.vuepress/theme/components/PageToc.vue:
```diff
@@ -0,0 +1,25 @@
+<template>
```
Seems that we lost item highlighting on navigation? Also, the navigation of the sections was being reflected in the URL, I'm not sure if this can be configured in the plugin, it'd be nice to preserve it.

Example:
![2022-06-14 at 11 57 53@2x](https://user-images.githubusercontent.com/83774467/173610016-707e4a5b-1f79-434a-b545-aee3dffacc82.png)

JonathanPaugh replied:
I will look into this, I think it was working at some point. It might have to do with the changes to the header levels.

braddialpad replied:
Yeah I remember seeing this working also, probably some type of customization that broke it

JonathanPaugh replied:
It seems that the problem is cause by not rendering the sub-headers in the sidebar. A hacky solution would be to use `display: none` on nested `ul` tags in the sidebar. I just found this [active-header-links plugin](https://v2.vuepress.vuejs.org/reference/plugin/active-header-links.html) which looks promising, so I will try it out first.

braddialpad replied:
I think that's actually a pretty decent solution if it works.

JonathanPaugh replied:
The plugin worked like a charm, should be all good now.

josedialpad commented on docs/.vuepress/theme/components/PageToc.vue:
```diff
@@ -0,0 +1,25 @@
+</script>
+
+<style lang="less">
+  .toc .toc-list {
```
The Dialtone class `d-ls-reset` can be used here to reset the default list style https://dialpad.design/utilities/typography/lists.html#resetting-a-list

josedialpad commented on docs/.vuepress/theme/components/PageToc.vue:
```diff
@@ -0,0 +1,25 @@
+<template>
+  <aside v-if="headers.length > 0" class="toc">
+    <h3 class="d-pl24 d-pb8 d-ws-nowrap">ON THIS PAGE</h3>
```
nit: add class `d-tt-uppercase` and use just the normal text: `On this page`. Not a big deal in this case, but here are [some reasons why doing this](https://medium.com/@mandy.michael/why-you-shouldnt-write-your-content-in-uppercase-instead-use-css-b03ac2c65b99).

josedialpad reviewed (commented):
Hey Jonathan, I'm leaving a couple of comments but overall great first contribution 🙌

braddialpad commented on docs/.vuepress/theme/index.js:
```diff
@@ -1,23 +1,25 @@
-const { path } = require('@vuepress/utils')
-const { defaultTheme } = require('vuepress')
-const { mediumZoomPlugin } = require('@vuepress/plugin-medium-zoom')
+const {path} = require('@vuepress/utils')
+const {defaultTheme} = require('vuepress')
+const {mediumZoomPlugin} = require('@vuepress/plugin-medium-zoom')
```
@josedialpad could you add a task for this in jira? Much more important now that we have vue code in the dialtone repo

josedialpad replied:
> Much more important now that we have vue code in the dialtone repo

I agree. Created https://dialpad.atlassian.net/browse/DT-509

JonathanPaugh commented:
Implemented all the mentioned changes.

braddialpad commented:
Super minor, but some of the links go from unwrapped to wrapped after selecting them such as "clipping text" here.

![Screen Shot 2022-06-14 at 12 03 54 PM](https://user-images.githubusercontent.com/64808812/173671746-e58d5740-db52-4b69-9cc2-0feba82ab478.png)
![Screen Shot 2022-06-14 at 12 04 08 PM](https://user-images.githubusercontent.com/64808812/173671764-4560e61a-0be4-450f-8bfe-bd7af93541a5.png)

josedialpad commented:
Another thing that I don't know if it's related to what Brad mentioned is that the highlight on different items is causing a small shift in the main content for example in the [breadcrumbs page](https://dialpad.design/deploy-previews/pr-615/components/breadcrumbs.html), clicking on "Usage" and then in "Variants and Examples" causes a left shift, then click on "Classes" and the content goes to the right again.

josedialpad commented:
> ## Obligatory GIF (super important!)
> ![](https://camo.githubusercontent.com/2d0eec3370009e359ba60dfec69094e5dea7e4a805b356fe4ee9cfeee1aa941f/68747470733a2f2f632e74656e6f722e636f6d2f71396d786d385f386c473441414141642f6769742d6d657267652d6769742e676966)

lol, I haven't seen the gif before 😂

braddialpad commented:
> lol, I haven't seen the gif before 😂

accurate, lol

braddialpad commented:
> clicking on "Usage" and then in "Variants and Examples" causes a left shift, then click on "Classes" and the content goes to the right again.

Yeah seems the whole right bar widens when you click variants and examples. Clicking through the rest of the others are fine. Maybe only happens on multi line items?

JonathanPaugh commented:
It seems to be only multi line headers that do that, seems to be okay if I use `d-ws-nowrap` on the item class. However I'm not sure if wrapping is something we want to keep. Imo it is probably better to have no wrapping, but long headers have the possibility of significantly reducing the content section size.

braddialpad commented:
> It seems to be only multi line headers that do that, seems to be okay if I use `d-ws-nowrap` on the item class. However I'm not sure if wrapping is something we want to keep. Imo it is probably better to have no wrapping, but long headers have the possibility of significantly reducing the content section size.

Can we set the TOC sidebar to a fixed width?

JonathanPaugh commented:
> Can we set the TOC sidebar to a fixed width?

Turns out it was cause by `font-weight` increasing the font size when the item is active, should be fixed now. Having a fixed size for the sidebar is possible, I don't know if it's needed now that this is working.

braddialpad reviewed (approved):
Lets wait until Julio's vuepress styling fix is in, merge that into this branch and make sure everything still works before merging.

braddialpad commented:
Julio's fix is in

braddialpad commented:
Looks like a big gap at the top of the page since the merge. Not sure if this was due to your changes or Julios.
![Uploading Screen Shot 2022-06-14 at 2.31.00 PM.png…]()

JonathanPaugh commented:
Yeah, just fixed that. I added padding to the top of the header but it seems like it is not needed anymore. I'm going to look around a bit more to see if there are any more styling problems.

braddialpad commented:
Some unintended changes on percy due to this.

braddialpad commented on packages/dialtone-css/lib/build/less/components/item-layout.less:
```diff
@@ -20,91 +20,56 @@
   line-height: var(--dt-font-line-height-300);
 
-  > .d-item-layout--content {
+  > .d-item-layout__content {
```
This naming was definitely wrong, but we'll have to be careful about breaking any existing CSS overrides in product

juliodialpad replied:
I was thinking a migration will still be needed anyway due to some changes on paddings, the overrides they were doing and to be able to remove the `unstyled` prop from the <item-layout>

I will take care of this migration, I can release this to a beta version and make the migration before merging this to staging.

braddialpad reviewed (commented):
Will test a bunch of example cases myself tomorrow. This looks great, but is also a bit of a scary change!

braddialpad commented on packages/dialtone-vue2/recipes/conversation_view/feed_item_row/feed_item_row.test.js:
```diff
@@ -126,7 +126,7 @@ describe('DtFeedItemRow tests', () => {
           ...basePropsData,
           showHeader: true,
-          avatarImageUrl: 'https://i1.sndcdn.com/avatars-000181324408-652e57-t500x500.jpg',
+          avatarImageUrl: '/avatar1.png',
```
Is this supposed to be an actual test image?

juliodialpad replied:
It does nothing on that test actually, just to avoid a warning maybe, but that file actually exists, the previous link wasn't working correctly tho.

braddialpad replied:
ah yeah, I should have commented the .story rather than the test. I just mean the image shows up like this:

![Screenshot 2025-02-03 at 10 47 21 AM](https://github.com/user-attachments/assets/3a79e90d-6f37-4605-aead-d5c396d747e0)

we should put in some sort of actual placeholder image.

juliodialpad replied:
Ah right, forgot about the base, will update to append the base URL

braddialpad reviewed (approved):
Did component comparison on the changed components. Looks good by me. Remember to do the migration, and also to do vue 3
